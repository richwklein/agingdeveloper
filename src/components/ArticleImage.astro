---
/**
 * @name ArticleImage
 * Component used for rendering the featured image for an article
 *
 * @param {Props} props - The properties of the featured image.
 * @return {Function} an Astro component factory function.
 *
 * @example
 * <ArticleImage featured={featured} />
 */
import type { ImageMetadata } from "astro";
import ExternalLink from "./ExternalLink.astro";
import { Picture } from "astro:assets";

interface Props {
  featured: {
    image: ImageMetadata;
    author?: {
      name: string;
      url: string;
    };
    site?: {
      name: string;
      url: string;
    };
    placeholder: string;
  };
  class?: string;
}
const {
  featured: { image, author, site, placeholder },
  class: className,
} = Astro.props;
const alt = author != null ? author.name : "unknown";
console.log(placeholder);
---

<section id="article-featured" class:list={[className]}>
  <figure>
    <Picture
      class:list={["aspect-video rounded-sm"]}
      src={image}
      alt={alt}
      formats={["webp"]}
      fallbackFormat={image.format}
      loading="eager"
      fetchpriority="high"
    />
    {
      author != null && site != null ? (
        <figcaption>
          <small class:list={["mt-1 text-sm text-primary-main/85"]}>
            {author != null ? (
              <>
                {"Image by "}
                <ExternalLink class:list={["underline"]} to={author.url}>
                  {author.name}
                </ExternalLink>
              </>
            ) : null}
            {site != null ? (
              <>
                {" on "}
                <ExternalLink class:list={["underline"]} to={site.url}>
                  {site.name}
                </ExternalLink>
              </>
            ) : null}
          </small>
        </figcaption>
      ) : null
    }
  </figure>
</section>
