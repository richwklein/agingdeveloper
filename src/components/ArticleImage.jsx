import React from "react";
import {GatsbyImage} from "gatsby-plugin-image";
import ExternalLink from "./ExternalLink";
import {Box, Typography} from "@mui/material";
import {ImageAuthorProps, ImageSiteProps} from "../props";
import PropTypes from "prop-types";

/**
 * React component for rendering an article's featured image.
 * The image should be in a Cinemascope 21:9 ratio.
 *
 * @param {ArticleImageProps} props - The article's featured image props.
 * @return {React.ReactElement} - The react component
 *
 * @example
 * <ArticleImage
 *   author={featured.author}
 *   site={featured.site}
 *   image={featured.image.childImageSharp.gatsbyImageData} />

 * @todo handle missing author, site, author.url, or site.url
 */
const ArticleImage = ({author, site, image}) => {
  return (
    <Box component="figure"
      sx={{
        my: 1.25,
        mx: 0,
      }}>
      <GatsbyImage image={image} />
      <Typography variant="caption" component="figcaption" sx={{mt: .25}}>
        {"Image by "}
        <ExternalLink to={author.url}>{author.name}</ExternalLink>
        {" on "}
        <ExternalLink to={site.url}>{site.name}</ExternalLink>
      </Typography>
    </Box>

  );
};

/**
 * @typedef ArticleImageProps - The article's featured image props.
 * @property {GatsbyImageData} image - The GatsbyImageData of the image.
 * @property {ImageAuthorProps} [author] - The author of the image.
 * @property {ImageSiteProps} [site] - The site the image comes from.
 */
ArticleImage.propTypes = {
  image: PropTypes.object.isRequired,
  author: ImageAuthorProps,
  site: ImageSiteProps,
};
export default ArticleImage;
