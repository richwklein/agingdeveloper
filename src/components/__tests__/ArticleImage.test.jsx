import React from "react";
import {cleanup, render} from "@testing-library/react";
import ArticleImage from "../ArticleImage";

// TODO improve tests
describe("ArticleImage", () => {
  afterEach(cleanup);

  const renderComponent = () => render(
      <ArticleImage
        author={{name: "Georgia Smith", url: "https://example.com/georgia-smith"}}
        site={{name: "Unsplash", url: "https://example.com/unsplash"}} />,
  );

  test("that it renders correctly.", () => {
    const {container} = renderComponent();
    expect(container.firstChild).toMatchSnapshot();
  });
});
