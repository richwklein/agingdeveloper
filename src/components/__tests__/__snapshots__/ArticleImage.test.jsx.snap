// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleImage that it renders correctly. 1`] = `
<figure
  class="MuiBox-root css-r5049d"
>
  <figcaption
    class="MuiTypography-root MuiTypography-caption css-12ijyxk-MuiTypography-root"
  >
    Image by 
    <a
      class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1xylxj1-MuiTypography-root-MuiLink-root"
      data-link="external"
      href="https://example.com/georgia-smith"
      rel="noopener noreferrer"
      target="_blank"
      to="https://example.com/georgia-smith"
    >
      Georgia Smith
    </a>
     on 
    <a
      class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1xylxj1-MuiTypography-root-MuiLink-root"
      data-link="external"
      href="https://example.com/unsplash"
      rel="noopener noreferrer"
      target="_blank"
      to="https://example.com/unsplash"
    >
      Unsplash
    </a>
  </figcaption>
</figure>
`;
